primitives/state-machine/src/lib.rs

```diff
@@ -726,6 +731,216 @@ mod execution {
 		prove_read_on_trie_backend(trie_backend, keys)
 	}
 
+	/// State machine only allows a single level
+	/// of child trie.
+	pub const MAX_NESTED_TRIE_DEPTH: usize = 2;
+
+	/// Multiple key value state.
+	/// States are ordered by root storage key.
+	#[derive(PartialEq, Eq, Clone)]
+	pub struct KeyValueStates(pub Vec<KeyValueState>);
+
+	/// A key value state.
+	#[derive(PartialEq, Eq, Clone)]
+	pub struct KeyValueState {
```

Comment:
Using name "State" here is a bit confusing. We already use the same term for the whole "State" as in the state of all storage for a block, including all subtrees. I'd suggest naming this something like "KeyValueStorageLevel"